### ISSUE TITLE
Implementing own AudioSink

### ISSUE THREAD

**Issue by ivan2311** [google/ExoPlayer#6180]
I would like to fix the **resampling** problem on Android, but still use `ExoPlayer` for playback. It seems to me that this could be fixed by avoiding using `DefaultAudioSink` implementation, that internally use `AudioTrack` for playback and have few `AudioProcessors` that can change sample rate of the output signal.
So, I need to implement my own `AudioSink` that wouldn't use `AudioTrack` or have those `AudioProcessors` that could change sample rate, but instead it would use some third party solution that doesn't have resampling problem.

Can you please help me in implementing my own `AudioSink`? Are there some sample projects with some custom implementation of `AudioSink`? What are the key points in the `AudioSink` implementation? Also, is it enough to just avoid using `AudioTrack`, and those few `AudioProcessors`, inside `AudioSink` implementation to prevent resampling on `ExoPlayer`?

**Comment by andrewlewis** [comment]
It would be useful if you could provide some more details about what you're trying to do:
- What is the input audio format and what format do you want to output?
- What is the 'resampling problem'?
- Do you want to play out the audio in the end?

If the answer to the last question is yes, it would probably make sense to use `DefaultAudioSink` but customize its behavior as needed.

**Comment by ivan2311** [comment]
Hi @andrewlewis ,

here are more details:

Q: _What is the input audio format and what format do you want to output?_
A: So, it's for the music streaming app and input format is Hi-Res FLAC. Sample rate could be up to 96 kHz and with 24bit bit depth. The output should have the same sample rate and bit depth.

Q: _What is the 'resampling problem'?_
A: As mentioned in previous answer, the output should have the same sample rate and bit depth, so basically it should be unchanged. The problem is that on Android, resampling happens, which I know it's not problem of the ExoPlayer library, but rather of the Android OS and how it handles Audio output. Resampling means that whatever the input is, the output will be resampled to devices native sample rate. E.g. lets say that we have a device with 48kHz native sample rate. When we play some FLAC file with 96kHz, the output file would be resampled to native 48kHz sample rate.

Q: _Do you want to play out the audio in the end?_
A: Yes, but problem with DefaultAudioSink is that it uses internally AudioTrack for playback. And with AudioTrack, resampling happens. Also, there are few AudioProcessors that changes bit depth, that I would also like to avoid.

**Comment by andrewlewis** [comment]
@ivan2311 How will you play out the 96 kHz audio without using `AudioTrack` and getting the resulting internal resampling behavior? Do you know of any devices with the hardware required to play this sample rate and an API that gives access to it to apps?

**Comment by andrewlewis** [comment]
I am also not sure that `AudioTrack` always resamples internally. The existence of the method [`isDirectPlaybackSupported`](https://developer.android.com/reference/android/media/AudioTrack#isDirectPlaybackSupported(android.media.AudioFormat,%2520android.media.AudioAttributes)) in Android Q suggests you may be able to get playback without resampling via `AudioTrack` on some devices.

**Comment by ivan2311** [comment]
@andrewlewis thanks for mentioning the existence of the method isDirectPlaybackSupported, I didn't know about it. It seems to me that there could be some newer devices, that will have Android Q, and could possibly support direct playback. But, currently I haven't read about any device that doesn't do resampling.

Also, I don't know where exactly resampling happens, is it inside AudioTrack or somewhere deeper inside Android Audio stack, but playing audio via AudioTrack will result with resampling to devices native sample rate, as far as I know. Even if you connect some external DAC to the Android device, resampling happens.

Currently, I'm investigating Superpowered library, since it can prevent resampling. But, I would still like to use ExoPlayer because of lots of other useful features. So, the plan is to use ExoPlayer, but instead of playing through AudioTrack, I would play audio through Superpowered library.

**Comment by andrewlewis** [comment]
You could also look at [AAudio](https://developer.android.com/ndk/guides/audio/aaudio/aaudio) and/or [Oboe](https://github.com/google/oboe), but, as you suggest, the resampling may be happening below a layer that you can control. For what it's worth, I am not aware of any ways to do audio output on Android apart from `AudioTrack`, OpenSL and AAudio (Oboe wraps these two).

Returning to your original question, the only documentation we provide on this is the javadoc on the `AudioSink` interface. I'd suggest also reading through the code for `DefaultAudioSink` and possibly using it as a starting point for your implementation. As the need to use a custom `AudioSink` is quite niche, I'm afraid we don't have time to write and maintain documentation on it, or to give detailed guidance here.

**Comment by ivan2311** [comment]
Yeah, I understand that it was a little bit too much to ask :) but nevertheless you still gave me some interesting info, so thank you!

As for the Oboe, I've already checked it and the resampling issue happens there also. Here is the issue with more info: [https://github.com/google/oboe/issues/526](https://github.com/google/oboe/issues/526)

**Comment by dturner** [comment]
Oboe author here, hello :) 

I'm keen to understand what you mean by "resampling issue". I understand that you want your application's audio bit depth and sample rate to match what comes out of the Android device (presumably via headphones, speaker or an attached USB audio device), however, those audio devices will have a fixed, non configurable (for now at least) bit depth and sample rate. 

This means that if your input audio format doesn't match that of the audio hardware's audio format resampling _must_ be done somewhere in the signal chain. It's not an _issue_ per se, it's a necessity.   

In Oboe if you request an audio stream without specifying a sample rate and bit depth you should receive the native sample rate and bit depth of the current default audio device. For example, if you had a 96kHz 24bit USB DAC attached you would receive those properties for your stream. 

What's your usage scenario? Are you looking for just high sample rate output, or low latency, or both, or something else - maybe attaching to an A/V receiver? 

Related: We're currently working on a resampler in Oboe which will allow you to request a stream with a given sample rate and Oboe will then do the resampling for you - allowing you to still obtain a low latency stream.

**Comment by ivan2311** [comment]
Hi @dturner ,

basically what I would like to achieve is written in this response: https://github.com/google/oboe/issues/526#issuecomment-496047221

and response to that was that it's not possible with Oboe. So, that's why I assumed that I can't achieve what I want with Oboe.

Also, here is the simple example of what I want to achieve:
I have Android device (LG V30) and external DAC (D50 Topping) with display that can show me the sample rate. When I play track with sample rate 48kHz with ExoPlayer, I want that same sample rate on the external DAC, also when I play track with sample rate 96kHz, I want same 96kHz on the external DAC. 
To be more exact, I have FLAC data on the input that is decoded to raw PCM data. The output data on the external DAC should be that same unchanged raw PCM data.

Can I achieve that with Oboe?

**Comment by ivan2311** [comment]
Also, if it's possible to achieve mentioned above with Oboe, I would still like to use ExoPlayer but maybe just use Oboe to do audio output, also if possible :)

**Comment by dturner** [comment]
There's 2 questions here: 

1) Can I configure the USB DAC to have a specific sample rate?
2) Can I obtain an audio stream with the same sample rate as the USB DAC?

I'll try to answer both below.

USB audio devices have one or more configurations for sample rate/bit depth. If you have a mac you can see the available configurations using the "Audio MIDI setup" app. Importantly, they can operate only in one configuration at time (which is what your D50 display is showing you). So for example, you cannot output a 96kHz audio file at the same time as outputting a 48kHz audio file at the same time without one of the audio streams being resampled at some point in the signal path. 

On some operating systems it's possible to change the configuration of the USB audio device after it's been connected, but on Android you cannot (currently, unless you package your own USB driver like [this app](https://play.google.com/store/apps/details?id=com.extreamsd.usbaudioplayerpro)). 

What Android does, at least AOSP Android (vendors, like Samsung or LG, might modify this behaviour) is query the available configurations when the USB device is connected and pick the configuration with the highest sample rate (and presumably bit depth as well). So in theory your D50, when connected to Android should have a native sample rate of 768kHz (what does it actually get on your LG V30)?

If it's configured at 768kHz, can you a stream with this sample rate? In theory, yes. You should be able to do this by requesting an audio stream in Oboe without specifying the sample rate or bit depth. Oboe will always try to give you a stream with the native sample rate. That said, 768kHz is a very high sample rate and there may be something else in the stream setup which limits the sample rate to something lower. Best thing to do is try it out - interested to know what happens in reality with this high end DAC. 

In short, you cannot configure your DAC from Android but you should be able to output at whatever the DAC is configured at without resampling.

**Comment by dturner** [comment]
Just realised the "output at whatever the DAC is configured at" statement is only true if you're using AAudio (recommended on Android API 27+). If you're using OpenSL ES it will try to open a stream at the same rate as the built-in audio device. 

What Android version do you have on your LG phone?

**Comment by ivan2311** [comment]
Yeah, LG V30 is on Android API 26 currently, so it's using OpenSL ES and it defaults to the same rate as the build-in audio device - 48 kHz. 

So, I've also tried with Huawei P20 Pro, API 28. It's fixed sample rate is 48kHz (used [PROPERTY_OUTPUT_SAMPLE_RATE](https://developer.android.com/reference/android/media/AudioManager#PROPERTY_OUTPUT_SAMPLE_RATE) to get it) and when connected to D50 DAC, it changes to 96kHz. Does that mean it uses sample rate to which the DAC is configured?
Also, when I fetch all sample rates for that D50 DAC (using [AudioDeviceInfo#getSampleRates()](https://developer.android.com/reference/android/media/AudioDeviceInfo#getSampleRates())) it returns this values: 44100, 48000, 88200, 96000, 176400, 192000. Shouldn't be using 192 kHz as highest sample rate of the connected DAC?

**Comment by dturner** [comment]
> It's fixed sample rate is 48kHz (used PROPERTY_OUTPUT_SAMPLE_RATE to get it) and when connected to D50 DAC, it changes to 96kHz. Does that mean it uses sample rate to which the DAC is configured?

Please be more specific rather than referring to "it". Do you mean the audio stream? the USB DAC or something else? 

> Shouldn't be using 192 kHz as highest sample rate of the connected DAC?

Well, like I said, there might be something else which limits the highest sample rate to 96kHz, maybe in Audio Policy Manager: https://source.android.com/devices/audio/implement-policy

**Comment by ivan2311** [comment]
> Please be more specific rather than referring to "it".

By it, I mean devices (Huawei P20 Pro) output audio stream. When it uses build-in speaker as audio output, sample rate is 48 kHz, and when it uses external USB DAC, sample rate is 96kHz.


> Well, like I said, there might be something else which limits the highest sample rate to 96kHz, maybe in Audio Policy Manager: https://source.android.com/devices/audio/implement-policy

👍

**Comment by dturner** [comment]
Thanks for clarifying. So the question is:

Is there any way of knowing what sample rate an external USB audio interface is configured at when not using AAudio? 

I'll need to do some digging on that.
